### ISSUE TITLE
Multiple log entries for 'action created' during migration

### ISSUE THREAD

**Issue by menakas** [woocommerce/action-scheduler#360]
On installing the latest beta, after migration, the scheduled actions page shows [multiple log entries for action created](https://cld.wthms.co/xXEprs). These logs show multiple entries for same time. This time matches the time of migration.

In the database, I can see same action id for these entries, but log id is different. So, it appears to be duplicate records created.

**Comment by rrennick** [comment]
Logging the action creation on migration should be disabled by https://github.com/Prospress/action-scheduler/blob/master/classes/migration/Runner.php#L98:L99 . We will need to investigate why it isn't.

@menakas If you create a new action (eg. add a webhook in WooCommerce), do you see 2 action created log entries?

**Comment by rrennick** [comment]
@menakas @james-allan I haven't been able to reproduce this in my install. Could you switch to using the `version_3_0_0` branch to see if the issue exists there?

**Comment by menakas** [comment]
@rrennick 
I used the [3.0.0 beta ](https://github.com/Prospress/action-scheduler/archive/3.0.0-beta-1.zip)to install. So, I think used the Version 3.0.0

**Comment by menakas** [comment]
I am able to replicate it. So, here are the steps I followed.
1. Spin up a test site.
1. Clone WooCommerce. Checkout v3.6.5 or v3.7.0( Can replicate with both)
1. Clone WooCommerce Subscriptions
1. Activate WC, WCS & set them up.
1. Create a few products and make a few purchases.
1. Install and activate [AS 3.0.0 beta](https://github.com/Prospress/action-scheduler/archive/3.0.0-beta-1.zip)
1. Now, look at the Scheduled Actions. 

I can find two duplicate entries for each Scheduled Action using the above method. I tried with WC v3.6.5 and WC v.3.7.0. In both cases, I can see this issue.

**Comment by thenbrent** [comment]
> I haven't been able to reproduce this in my install

@rrennick do you have have WC Admin active and/or some other custom data store? If so, perhaps that is why it's not happening on your site as that changes the logic for the default data store / logger.

**Comment by james-allan** [comment]
> Could you switch to using the version_3_0_0 branch to see if the issue exists there?

I used the `version_3_0_0` branch and got the same result. 

To debug further I've logged the stack traces for each 'action created' entry and what I found is that its same trace every time. I've included it below: 

```
#0 wp-includes/class-wp-hook.php(310): WP_Hook->apply_filters('', array)
#1 wp-includes/plugin.php(465): WP_Hook->do_action(array)
#2 action-scheduler-version_3_0_0/classes/data-stores/ActionScheduler_DBStore.php(53): do_action('action_scheduler_stored_action', 10469)
#3 action-scheduler-version_3_0_0/classes/migration/ActionScheduler_DBStoreMigrator.php(31): ActionScheduler_DBStore->save_action(object(ActionScheduler_Action), NULL)
#4 action-scheduler-version_3_0_0/classes/data-stores/ActionScheduler_HybridStore.php(225): ActionScheduler_DBStoreMigrator->save_action(object(ActionScheduler_Action), NULL)
#5 action-scheduler-version_3_0_0/classes/ActionScheduler_ActionFactory.php(177): ActionScheduler_HybridStore->save_action(object(ActionScheduler_Action))
#6 action-scheduler-version_3_0_0/classes/ActionScheduler_ActionFactory.php(84): ActionScheduler_ActionFactory->store(object(ActionScheduler_Action))
#7 action-scheduler-version_3_0_0/functions.php(30): ActionScheduler_ActionFactory->single('action_scheduler/migration_hook', array, 1569812198, 'action-scheduler-migration')
#8 action-scheduler-version_3_0_0/classes/migration/Scheduler.php(89): as_schedule_single_action(1569812198, 'action_scheduler/migration_hook', array, 'action-scheduler-migration')
#9 action-scheduler-version_3_0_0/classes/migration/Scheduler.php(46): Action_Scheduler\Migration\Scheduler->schedule_migration(1569812198)
#10 wp-includes/class-wp-hook.php(284): Action_Scheduler\Migration\Scheduler->run_migration()
#11 wp-includes/class-wp-hook.php(310): WP_Hook->apply_filters('', array)
#12 wp-includes/plugin.php(531): WP_Hook->do_action(array)
#13 action-scheduler-version_3_0_0/classes/actions/ActionScheduler_Action.php(22): do_action_ref_array('action_scheduler/migration_hook', array)
#14 action-scheduler-version_3_0_0/classes/abstracts/ActionScheduler_Abstract_QueueRunner.php(61): ActionScheduler_Action->execute()
#15 action-scheduler-version_3_0_0/classes/ActionScheduler_QueueRunner.php(150): ActionScheduler_Abstract_QueueRunner->process_action(10218, 'Async Request')
#16 action-scheduler-version_3_0_0/classes/ActionScheduler_QueueRunner.php(120): ActionScheduler_QueueRunner->do_batch(25, 'Async Request')
#17 wp-includes/class-wp-hook.php(286): ActionScheduler_QueueRunner->run('Async Request')
#18 wp-includes/class-wp-hook.php(310): WP_Hook->apply_filters('', array)
#19 wp-includes/plugin.php(465): WP_Hook->do_action(array)
#20 action-scheduler-version_3_0_0/classes/ActionScheduler_AsyncRequest_QueueRunner.php(52): do_action('action_scheduler_run_queue', 'Async Request')
#21 woocommerce/includes/libraries/wp-async-request.php(147): ActionScheduler_AsyncRequest_QueueRunner->handle()
#22 wp-includes/class-wp-hook.php(286): WP_Async_Request->maybe_handle('')
#23 wp-includes/class-wp-hook.php(310): WP_Hook->apply_filters('', array)
#24 wp-includes/plugin.php(465): WP_Hook->do_action(array)
#25 wp-admin/admin-ajax.php(173): do_action('wp_ajax_as_async_request_queue_runner')
```

IDK what this means yet. I'm going to start digging into now. I have a feeling the async runner is claiming the same scheduled actions for migration or something? 🤷‍♂

**Comment by james-allan** [comment]
> IDK what this means yet. 

ok I think this is fine. What I think is happening is that there are multiple logger instances hooked in and everyone of them is logging the note. This goes back to what @rrennick said earlier:

> Logging the action creation on migration should be disabled by https://github.com/Prospress/action-scheduler/blob/master/classes/migration/Runner.php#L98:L99 . We will need to investigate why it isn't.

I've had a look at this and when the function that unhooks the logger is ran, sometimes there are still 4 other instances of `ActionScheduler_DBLogger::log_stored_action()` hooked on and they have slightly different "IDs":

For example this is the output of the `wp_filter` global:
```
array(
   'callbacks' => 
  array (
    10 => 
    array (
      '0000000041bb79b0000000005ba5231dlog_stored_action' => 
      array (
        'function' => 
        array (
          0 => 
          ActionScheduler_DBLogger::__set_state(array(
          )),
          1 => 'log_stored_action',
        ),
        'accepted_args' => 1,
      ),
      '0000000041bb057e000000005ba5231dlog_stored_action' => 
      array (
        'function' => 
        array (
          0 => 
          ActionScheduler_DBLogger::__set_state(array(
          )),
          1 => 'log_stored_action',
        ),
        'accepted_args' => 1,
      ),
    ),
  ),
 ...
))
```

In this example there are still 2 other instances hooked on

`0000000041bb79b0000000005ba5231dlog_stored_action`
`0000000041bb057e000000005ba5231dlog_stored_action`
these "IDs" are generated by [`_wp_filter_build_unique_id()`](https://developer.wordpress.org/reference/functions/_wp_filter_build_unique_id/).

So we just need to find: 

1. Why there are multiple instances of the function still hooked on?
1. What we can do to fix it?
1. Are there alternatives, like using a static function instead?

**Comment by james-allan** [comment]
A little more information. I added the `_wp_filter_build_unique_id()` key to the log entry and this is what I saw: 

<img width="972" alt="Screen Shot 2019-09-30 at 3 40 47 pm" src="https://user-images.githubusercontent.com/8490476/65852181-b607a800-e398-11e9-9070-51ba41c9e6eb.png">

It's hard to tell the difference at a glance but they are all different and they are all destination logger objects. 

So somewhere in the background there are 5 destination logger objects all logging the note. 

After looking into where these 5 destination loggers could be coming from I found that they are instantiated in [`Controller::get_migration_config_object()`](https://github.com/Prospress/action-scheduler/blob/52bd35581c73c77094b0b2dd4b639e595a009457/classes/migration/Controller.php#L115) which is called by `Scheduler::get_migration_runner()` which is called from `Scheduler::run_migration()`.

That function is hooked onto the `action_scheduler/migration_hook` action which can run multiple times in the same request.

To confirm this, I stored the destination logger IDs in a global and then logged them on shutdown and this is what that looked like: 

```
Array
(
    [0] => 0000000072721f99000000007117f2a1
    [1] => 0000000072721393000000007117f2a1
    [2] => 00000000727268a4000000007117f2a1
    [3] => 000000007272c407000000007117f2a1
    [4] => 0000000072735560000000007117f2a1
    [5] => 000000007273a4ef000000007117f2a1
    [6] => 0000000072702a74000000007117f2a1
)

```

So to fix this I think we need to make the config object a singleton so a new instance isn't created every time. What do you think @rrennick? Do you see any issues with that approach?

**Comment by thenbrent** [comment]
> That function is hooked onto the action_scheduler/migration_hook action which can run multiple times in the same request.

The reason we'd be seeing that now, and not earlier on, is because of https://github.com/Prospress/action-scheduler/commit/a90dcaced1ebfebfa83b979e945886930e9254f0

**Comment by rrennick** [comment]
@menaka @james-allan Thanks for the additional research. This should be fixed with #360.

**Comment by thenbrent** [comment]
Fixed with #366.
